examples/files.rs

```diff
@@ -0,0 +1,218 @@
+use std::collections::HashSet;
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+use std::time::{Duration, Instant};
+
+use crossterm::event::{Event, KeyCode, MouseEventKind};
+use ratatui::backend::{Backend, CrosstermBackend};
+use ratatui::layout::Rect;
+use ratatui::style::{Color, Modifier, Style};
+use ratatui::text::{Line, Span};
+use ratatui::widgets::{Block, Scrollbar, ScrollbarOrientation};
+use ratatui::{Frame, Terminal};
+use tui_tree_widget::{Node, Tree, TreeData, TreeState};
+
+struct FileTreeData(PathBuf);
+
+impl TreeData for FileTreeData {
+    type Identifier = OsString;
+
+    fn flatten(&self, open: &HashSet<Vec<Self::Identifier>>) -> Vec<Node<Self::Identifier>> {
+        let mut result = Vec::new();
+        flatten_inner(&mut result, open, &self.0, &[]);
+        result
+    }
+
+    fn render(
+        &self,
+        identifier: &[Self::Identifier],
+        area: ratatui::layout::Rect,
+        buffer: &mut ratatui::buffer::Buffer,
+    ) {
+        let mut path = self.0.clone();
+        for i in identifier {
+            path = path.join(i);
+        }
+
+        let mut spans = Vec::new();
+
+        if let Some(filename) = path.file_stem() {
+            spans.push(Span::raw(filename.to_string_lossy()));
+        }
+
+        if let Some(extension) = path.extension() {
+            const STYLE: Style = Style::new().fg(Color::DarkGray);
+            spans.push(Span::styled(".", STYLE));
+            spans.push(Span::styled(extension.to_string_lossy(), STYLE));
+        }
+
+        ratatui::widgets::Widget::render(Line::from(spans), area, buffer);
+    }
+}
+
+fn flatten_inner(
+    result: &mut Vec<Node<OsString>>,
+    open: &HashSet<Vec<OsString>>,
+    path: &Path,
+    current: &[OsString],
+) {
+    let Ok(read_dir) = path.read_dir() else {
```

Comment:
```suggestion
    let Ok(entries) = path.read_dir() else {
```

Do you think it could be a good idea to handle the errors by displaying them in the tree instead? E.g.:

```rust
let Ok(entries) = path.read_dir() else {
    result.push(someErrorNode);
    return;
}
for entry in entries {
    let Ok(entry) = entry else {
       result.push(someErrorNode);
       continue;
    }
    ...
}
```

Reply:
It annoys me more to have no sorting here but I think this should serve as a base example on how to generally approach this.

Reply:
Variable name: the returned struct is named ReadDir which is why I named it that way.
https://doc.rust-lang.org/std/path/struct.Path.html#method.read_dir

Reply:
I mentioned this because not being hugely familiar with the read_dir function, I had to look up what it returns in order to understand the code fully. Naming it entries directly would have made that . `ReadDir` is an `iterator over the entries in a directory.`